### ISSUE TITLE
Migrating from an OAuth1 Private App

### ISSUE THREAD

**Issue by gierschv** [XeroAPI/xero-node#331]
Hi there,

I discovered last week when I tried to create a new OAuth1 private dev app that this was no longer possible, and the OAuth1 will be deprecated in December 2020. I didn't receive any communications for that, especially since we have apps in production for our UK and US Xero accounts...

Since I no longer have a way to test our integration with a dev/test environment (but we have it in production, yayyy!), I started looking at migrating to xero-node 4.x and the OAuth2 auth.

My use case is pretty basic, we use the API to create invoices for our customers, managing contacts, create quotes (for now using draft invoices since we didn't had a Quote API), get/push files linked to invoices, manage transactions, reconcile with our different accounts, etc.

I was able to run [your example](https://github.com/XeroAPI/xero-node#javascript-example) and get a refresh token to put in our config file (since we only have two tenants), but I didn't find a proper way yet to instantiate the client and get a access token from the refresh token yet.

It seems that the new client was designed to only use the API with an interactive authentication at every usage. The example seems to be designed with some multi-tenants in mind (as a public app?), but there is a single client instance and the tokens are stored in the client itself... This doesn't really make sense at all. Is this v4 client at alpha or design stage?

I was looking for a solution in the issues, and also saw [this issue #304](https://github.com/XeroAPI/xero-node/issues/304): I was wondering where can I find the TS code for the 4.0.7 release? The master branch is behind and the last release on GitHub as 4.0.4, and I only have the transpiled code in the NPM build, which has additional methods than master (e.g. `buildClient()`).

Thanks

EDIT: Found a way, but...

```js
const xero_node = require('xero-node');
const TokenSet = require('openid-client').TokenSet;

xero = new xero_node.XeroClient({
  clientId: config.xero.GB.client_id,
  clientSecret: config.xero.GB.client_secret,
  scopes: ['openid', 'profile', 'email', 'accounting.transactions', 'accounting.contacts', 'accounting.attachments', 'files', 'offline_access'],
});
await xero.buildClient();
xero.setTokenSet(new TokenSet({
  refresh_token: config.xero.GB.refresh_token,
  access_token: '',
}));
await xero.refreshToken();

console.log(xero.accountingApi);
console.log(await xero.accountingApi.getOrganisations(config.xero.GB.tenant_id));
```

But I just noticed that the new OAuth2 authentication is even more terrible. Once you use a refresh token, this one is instantly invalidated... 

![Screenshot 2020-01-20 at 13 50 11](https://user-images.githubusercontent.com/396537/72728137-f4fe7b00-3b8c-11ea-8436-b11a7c8b1829.png)

Now I guess I need to have a DB/centralized config storage to keep the last access token / refresh token, and avoid losing the response of the last refresh request... otherwise I need to manually re-authorize it. 

How we can use the API and new client in the real world? I'm not even talking about running our functional tests or put these tests in a CI.

**Comment by danielgreane** [comment]
Hi @gierschv ,

Sorry you're in the same boat. I've emailed Xero support to ask if we can continue to issue OAuth1 tokens, but they've refused, so imagine you're in a similar boat to us where you can't develop any new features until you upgrade because you can't generate an OAuth1 sandbox account. I get that they're deprecating at the end of this year and want to encourage upgrades... but 16 months in advance with no option to upgrade slowly seems insane

We're thinking along the same lines of having to:

1. Generate a token with a manual step, and saving this to a database (even though this is terrible practice from a security point of view)
2. Run a task that refreshes the token every X minutes and updates the record in the database
3. Read the token from the DB before an API call is made... 

Unless we're both missing something here, this looks like a nasty work-around due to Xero's new OAuth2 flow not really being built with async programatic access in mind

**Comment by wobinb** [comment]
@danielgreane @gierschv We are able to allow you to create further OAuth 1 Private apps for the time being, can I ask you to email our support team using api@xero.com Once you've done that please let me know your CX reference number so that we can email you the details.

**Comment by danielgreane** [comment]
Thanks @wobinb, firing over an email now however the issue is that your sandbox / demo accounts delete themselves every 30 days. Will we have to follow this process and request a new OAuth1 app every 30 days?

Of course eventually we will need to migrate to OAuth2, however our use cases above is a pretty common one... is something in the roadmap to allow our use cases above? Is this merely a limitation in the node SDK, or is it a design limitation of Xero's implementation of OAuth2?

**Comment by danielgreane** [comment]
CX0009432331

> @danielgreane @gierschv We are able to allow you to create further OAuth 1 Private apps for the time being, can I ask you to email our support team using [api@xero.com](mailto:api@xero.com) Once you've done that please let me know your CX reference number so that we can email you the details.

**Comment by gierschv** [comment]
@wobinb CX0009432361, thank you!

@danielgreane I found this thread today (https://community.xero.com/developer/discussion/109207632): they are trying to deprecate private apps. And we got the Node SDK that doesn't make that easy either by not handling this use case.

**Comment by SidneyAllen** [comment]
@gierschv and @danielgreane - we'll be devoting time this week to work through the backlog of issues on the node sdk.   The mismatch between the master and 4.07 on NPM was a PR that hadn't been merged.  We'll work harder to keep our releases in sync with NPM.

**Comment by danielgreane** [comment]
Thanks @SidneyAllen , I've been setup with a legacy app with support in the meantime so we're not paralysed with feature releases. I'll keep an eye out and hopefully have ample time to get migrated to the new API in a few weeks when there's some clarity around how to handle M2M interactions and refresh the token in a predictable manner

**Comment by SerKnight** [comment]
Hey @gierschv 
> ...but I didn't find a proper way yet to instantiate the client and get a access token from the refresh token yet.

So to expand on the usage of this SDK we've created a sample node app with basic usage of most endpoints ( https://github.com/XeroAPI/xero-node-oauth2-app ) - a few still in the works.

The OAuth2.0 migration is a big change from our previous architecture so I completely understand the frustration.

On the topic of 'forever access' as we once had it - global data security standards have forced us to re-evaluate how we provision API access. While Private apps will continue to be supported through the end of year - the access token / refresh token strategy is going to be a safer data security architecture for both our customers & the users of our communities products.

I hope that makes sense and happy to discus more. We've got some more dev evangelists on the team so support for SDK's will be noticeably increased over the next quarter! 🙇 
-----

We just released 4.1.0 and tagged a release with the detailed CHANGELOG - the sample app will also show a strategy for persisting tokens in session, changing tenants, and refreshing the token.

Closing this issue for now (4.0.7 was my mistake waiting on a PR merge) but if you get sample app running and still confused please re-open.

Sample App:
<img width="1426" alt="Screen Shot 2020-01-21 at 5 32 33 PM" src="https://user-images.githubusercontent.com/2935387/72855342-5e0cee00-3c74-11ea-8942-eeabc2a51ba3.png">
